server/connector.go

```diff
@@ -222,10 +242,26 @@ func (c *Connector) HandleConnection(ctx context.Context, frontendConn net.Conn)
 			WithField("handshake", handshake).
 			Debug("Got handshake")
 
-		serverAddress := handshake.ServerAddress
-		nextState := mcproto.State(handshake.NextState)
+		if handshake.NextState == mcproto.StateLogin {
+			userInfo, err := c.readUserInfo(bufferedReader, clientAddr, handshake.NextState)
+			if err != nil {
+				logrus.
+					WithError(err).
+					WithField("clientAddr", clientAddr).
+					Error("Failed to read user info")
+				c.metrics.Errors.With("type", "read").Add(1)
+				return
+			}
+			logrus.
+				WithField("client", clientAddr).
+				WithField("userInfo", userInfo).
+				Debug("Got user info")
+
+			c.findAndConnectBackend(ctx, frontendConn, clientAddr, inspectionBuffer, handshake.ServerAddress, userInfo, handshake.NextState)
+		} else {
+			c.findAndConnectBackend(ctx, frontendConn, clientAddr, inspectionBuffer, handshake.ServerAddress, nil, handshake.NextState)
```

Comment:
You could just pass an empty user information object here and avoid calling the same function twice here.

Reply:
It needs to be nil for the JSON field omission to work as I want. With that said, I can shift it up to a var to avoid the ugly double-call.